nion/swift/ExportDialog.py

```diff
@@ -246,66 +250,176 @@ def cancel(self) -> bool:
-    def __init__(self, document_controller: DocumentController.DocumentController, display_item: DisplayItem.DisplayItem) -> None:
+    def __init__(self, document_controller: DocumentController.DocumentController,
+                 display_item: DisplayItem.DisplayItem) -> None:
         super().__init__()
-
         self.__document_controller = document_controller
+        self.__display_item = display_item
+        self.__model = ExportSizeModel(display_item)
+        self.__handler = ExportSVGHandler(self.__model)
+        self.__init_ui()
 
+    def __init_ui(self) -> None:
         u = Declarative.DeclarativeUI()
+        dialog = typing.cast(Dialog.ActionDialog, Declarative.construct(
+            self.__document_controller.ui,
+            self.__document_controller,
+            u.create_modeless_dialog(
+                self.__handler.ui_view, title=_("Export SVG")
+            ),
+            self.__handler
+        ))
+        dialog.add_button(_("Cancel"), self.__cancel_clicked)
+        dialog.add_button(_("Export"), self.__ok_clicked)
+        dialog.show()
 
-        if display_item.display_data_shape and len(display_item.display_data_shape) == 2:
-            display_size = Geometry.IntSize(height=4, width=4)
-        else:
-            display_size = Geometry.IntSize(height=3, width=4)
-
-        handler = ExportSVGHandler(display_size)
-
-        def ok_clicked() -> bool:
-            dpi = 96
-            width_px = (handler.width_model.value or display_size.width) * dpi
-            height_px = (handler.height_model.value or display_size.height) * dpi
-
-            ui = document_controller.ui
-            filter = "SVG File (*.svg);;All Files (*.*)"
-            export_dir = ui.get_persistent_string("export_directory", ui.get_document_location())
-            export_dir = os.path.join(export_dir, display_item.displayed_title)
-            path, selected_filter, selected_directory = document_controller.get_save_file_path(_("Export File"), export_dir, filter, None)
-            if path and not os.path.splitext(path)[1]:
-                path = path + os.path.extsep + "svg"
-            if path:
-                ui.set_persistent_string("export_directory", selected_directory)
-                display_shape = Geometry.IntSize(height=height_px, width=width_px)
-                document_controller.export_svg_file(DisplayPanel.DisplayPanelUISettings(ui), display_item, display_shape, pathlib.Path(path))
-            return True
-
-        def cancel_clicked() -> bool:
-            return True
-
-        dialog = typing.cast(Dialog.ActionDialog, Declarative.construct(document_controller.ui, document_controller, u.create_modeless_dialog(handler.ui_view, title=_("Export SVG")), handler))
-        dialog.add_button(_("Cancel"), cancel_clicked)
-        dialog.add_button(_("Export"), ok_clicked)
+    def __ok_clicked(self) -> bool:
+        display_shape = Geometry.IntSize(width=int(self.__model.width), height=int(self.__model.height))
```

Comment:
This is wrong. You should add `display_size_in_pixels` to the model and use that property here. Instead, you're just grabbing whatever the last field values are, ignoring the units. I would also suggest renaming this local variable `display_shape` to `display_shape_in_pixels` so that it is clear.

Reply:
Changed to in model public function